sklearn/utils/estimator_checks.py

```diff
@@ -3121,6 +3121,83 @@ def check_requires_y_none(name, estimator_orig, strict_mode=True):
             warnings.warn(warning_msg, FutureWarning)
 
 
+def check_dataframe_column_names_consistency(name, estimator_orig):
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest("pandas is not installed: not checking "
+                       "column name consistency for pandas")
+
+    def _construct_dataframe(X, columns):
+        return pd.DataFrame(X, columns=columns)
+    _check_column_name_consistency(name, estimator_orig, _construct_dataframe,
+                                   "dataframe")
+
+
+def check_dataarray_column_names_consistency(name, estimator_orig):
+    try:
+        import xarray as xr
+    except ImportError:
+        raise SkipTest("xarray is not installed: not checking "
+                       "column name consistency for xarray")
+
+    def _construct_xarray(X, columns):
+        return xr.DataArray(X, dims=('index', 'columns'),
+                            coords={'columns': columns})
+    _check_column_name_consistency(name, estimator_orig, _construct_xarray,
+                                   "xarray")
+
+
+def _check_column_name_consistency(name, estimator_orig, construct_X,
+                                   array_name):
+    estimator = clone(estimator_orig)
+    tags = estimator._get_tags()
+    if "2darray" not in tags["X_types"] or tags["no_validation"]:
+        return
+
+    X_orig, _ = make_regression(random_state=0, n_features=5)
+    X_orig = _enforce_estimator_tags_x(estimator, X_orig)
+    X_orig = _pairwise_estimator_convert_X(X_orig, estimator)
+
+    n_samples, n_features = X_orig.shape
+    names = np.array([f"col_{i}" for i in range(n_features)])
+    X = construct_X(X_orig, names)
+
+    rng = np.random.RandomState(0)
+    y = rng.randint(low=0, high=2, size=n_samples)
+    y = _enforce_estimator_tags_y(estimator, y)
+
+    estimator.fit(X, y)
+
+    if not hasattr(estimator, "feature_names_in_"):
+        raise ValueError("Estimator does not have a feature_names_in_ "
+                         f"attribute after fitting with a {array_name}")
+
+    assert_array_equal(estimator.feature_names_in_, names)
+    bad_names = names[::-1]
```

Comment:
Would it make also sense to have a smoke test by running the methods with `names` to be sure that no error is raised?